pallets/processor-manager/src/lib.rs

```diff
@@ -333,5 +364,77 @@ pub mod pallet {
 
             Ok(().into())
         }
+
+        #[pallet::call_index(5)]
+        #[pallet::weight(T::WeightInfo::heartbeat_with_version())]
+        pub fn heartbeat_with_version(
+            origin: OriginFor<T>,
+            version: Version,
+        ) -> DispatchResultWithPostInfo {
+            let who = ensure_signed(origin)?;
+            _ = Self::manager_id_for_processor(&who).ok_or(Error::<T>::ProcessorHasNoManager)?;
+
+            <ProcessorHeartbeat<T>>::insert(&who, T::UnixTime::now().as_millis());
+            <ProcessorVersion<T>>::insert(&who, version.clone());
+
+            Self::deposit_event(Event::<T>::ProcessorHeartbeatWithVersion(who, version));
+
+            Ok(().into())
+        }
+
+        #[pallet::call_index(6)]
+        #[pallet::weight(T::WeightInfo::insert_binary_hash())]
+        pub fn insert_binary_hash(
+            origin: OriginFor<T>,
+            version: Version,
+            hash: BinaryHash,
+        ) -> DispatchResultWithPostInfo {
+            ensure_root(origin)?;
+
+            <KnownBinaryHash<T>>::insert(&version, hash.clone());
+
+            Self::deposit_event(Event::<T>::BinaryHashInserted(version, hash));
+
+            Ok(().into())
+        }
+
+        #[pallet::call_index(7)]
+        #[pallet::weight(T::WeightInfo::remove_binary_hash())]
+        pub fn remove_binary_hash(
+            origin: OriginFor<T>,
+            version: Version,
+        ) -> DispatchResultWithPostInfo {
+            ensure_root(origin)?;
+
+            let hash = <KnownBinaryHash<T>>::take(&version);
+
+            if let Some(hash) = hash {
+                Self::deposit_event(Event::<T>::BinaryHashRemoved(version, hash));
+            }
+
+            Ok(().into())
+        }
```

Comment:
This could be merged into a single call `update_binary_hash`, where `hash: BinaryHash` would be an option.